Adding a useful topic of how a custom Import Entity can be defined. (#5007)
magento/devdocs: guides/v2.2/ext-best-practices/tutorials/custom-import-entity.md

---
group: extension-best-practices
title: Custom import entity
contributor_name: Atwix
contributor_link: https://www.atwix.com/
---

This tutorial shows you how to add a new import entity, that can be used to import data into your custom module's table, by extending the [Magento/ImportExport/Model/Import/Entity/AbstractEntity][0]{:target="_blank"} class. 
The current import entities can be found in **System** > **Import**:

- Advanced Pricing
- Products
- Customers and Addresses (single file)
- Customers Main File
- Customer Addresses

> For the beginning let's suppose that we have a custom table with the following structure:

Reviewer: ```suggestion
To begin, let's suppose we have a custom table with the following structure:
```

| entity_id | name | duration |
| --- | --- | --- |
|           |  |  |

> ## Step 1: Adding a new Entity Type

Reviewer: ```suggestion
## Step 1: Adding a New Entity Type
```

Declaring our new import entity:

> `etc/import.xml`

```xml
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_ImportExport:etc/import.xsd">
    <entity name="learning" label="Learning Courses Import" model="OrangeCompany\Learning\Model\Import\Courses"
            behaviorModel="Magento\ImportExport\Model\Source\Import\Behavior\Basic" />
</config>
```

> As we extend the **Magento_ImportExport** module, we should also add a dependency to it in `module.xml` file.

Reviewer: ```suggestion
As we extend the **Magento_ImportExport** module, we should also add a dependency to it in the `module.xml` file.
```

                $this->connection->insertOnDuplicate($tableName, $rows, $this->getAvailableColumns());

                return true;
            }

            return false;
        }
    }

    /**
     * Delete entities
     *
     * @param array $entityIds
     *
     * @return bool
     */
    private function deleteEntityFinish(array $entityIds): bool
    {
        if ($entityIds) {
            try {
                $this->countItemsDeleted += $this->connection->delete(
                    $this->connection->getTableName(static::TABLE),
                    $this->connection->quoteInto(static::ENTITY_ID_COLUMN . ' IN (?)', $entityIds)
                );

                return true;
            } catch (Exception $e) {
                return false;
            }
        }

        return false;
    }

    /**
     * Get available columns
     *
     * @return array
     */
    private function getAvailableColumns(): array
    {
        return $this->validColumnNames;
    }
}
```

{% endcollapsible %}

## Result

As result, we should be able to see the new Entity Type:
 
![Import Entity]({{ site.baseurl }}/common/images/ext-best-practices/import-entity.png)
 

>  Here is a sample csv file that can be used for importing data into our table.

Reviewer: ```suggestion
 Here is a sample csv file we can use to import data into our table.
```

```text
entity_id,name,duration
,"First Course",90
,"Second Course",120
```

{:.bs-callout .bs-callout-info}

> For updating the table's data, the `entity_id` value is required to be provided for the respectively row.

Reviewer: ```suggestion
For updating the table's data, you must provide the `entity_id` value for each row.
```
